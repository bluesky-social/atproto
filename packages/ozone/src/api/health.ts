import express from 'express'
import { sql } from 'kysely'
import AppContext from '../context'

export const createRouter = (ctx: AppContext): express.Router => {
  const router = express.Router()

  router.get('/robots.txt', function (req, res) {
    res.type('text/plain')
    res.send(
      '# Hello Friends!\n\n# Crawling the public parts of the API is allowed. HTTP 429 ("backoff") status codes are used for rate-limiting. Up to a handful concurrent requests should be ok.\nUser-agent: *\nAllow: /',
    )
  })

  router.get('/xrpc/_health', async function (req, res) {
    const { version } = ctx.cfg.service
    try {
      await sql`select 1`.execute(ctx.db.db)
    } catch (err) {
      req.log.error(err, 'failed health check')
      return res.status(503).send({ version, error: 'Service Unavailable' })
    }
    res.send({ version })
  })

  return router
}
