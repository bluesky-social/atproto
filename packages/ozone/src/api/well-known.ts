import express from 'express'
import AppContext from '../context'

export const createRouter = (ctx: AppContext): express.Router => {
  const router = express.Router()

  router.get('/.well-known/did.json', (_req, res) => {
    const hostname =
      ctx.cfg.service.publicUrl && new URL(ctx.cfg.service.publicUrl).hostname
    if (!hostname || ctx.cfg.service.did !== `did:web:${hostname}`) {
      return res.sendStatus(404)
    }
    res.json({
      '@context': ['https://www.w3.org/ns/did/v1'],
      id: ctx.cfg.service.did,
      verificationMethod: [
        {
          id: `${ctx.cfg.service.did}#atproto_label`,
          type: 'Multikey',
          controller: ctx.cfg.service.did,
          publicKeyMultibase: ctx.signingKey.did().replace('did:key:', ''),
        },
      ],
      service: [
        {
          id: '#atproto_labeler',
          type: 'AtprotoLabeler',
          serviceEndpoint: `https://${hostname}`,
        },
      ],
    })
  })

  return router
}
