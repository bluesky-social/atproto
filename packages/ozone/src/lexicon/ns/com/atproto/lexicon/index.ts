/**
 * GENERATED CODE - DO NOT MODIFY
 */
import {
  type Auth,
  type Options as XrpcOptions,
  Server as XrpcServer,
  type StreamConfigOrHandler,
  type MethodConfigOrHandler,
  createServer as createXrpcServer,
} from '@atproto/xrpc-server'
import * as ComAtprotoLexiconSchema from '../../../../types/com/atproto/lexicon/schema.js'
import { Server } from '../../../../index.js'

export class ComAtprotoLexiconNS {
  _server: Server

  constructor(server: Server) {
    this._server = server
  }
}
