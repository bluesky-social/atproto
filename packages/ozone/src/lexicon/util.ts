/**
 * GENERATED CODE - DO NOT MODIFY
 */
export function isObj(v: unknown): v is Record<string, unknown> {
  return typeof v === 'object' && v !== null
}

export function hasProp<K extends PropertyKey>(
  data: object,
  prop: K,
): data is Record<K, unknown> {
  return prop in data
}
