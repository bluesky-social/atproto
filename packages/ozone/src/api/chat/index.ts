import { Server } from '../../lexicon'
import AppContext from '../../context'
import getActorMetadata from './getActorMetadata'
import getMessageContext from './getMessageContext'

export default function (server: Server, ctx: AppContext) {
  getActorMetadata(server, ctx)
  getMessageContext(server, ctx)
  return server
}
