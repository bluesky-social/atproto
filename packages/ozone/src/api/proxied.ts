import { Server } from '../lexicon'
import AppContext from '../context'
import { ids } from '../lexicon/lexicons'

export default function (server: Server, ctx: AppContext) {
  server.app.bsky.actor.getProfile({
    auth: ctx.authVerifier.moderator,
    handler: async (request) => {
      const res = await ctx.appviewAgent.api.app.bsky.actor.getProfile(
        request.params,
        await ctx.appviewAuth(ids.AppBskyActorGetProfile),
      )
      return {
        encoding: 'application/json',
        body: res.data,
      }
    },
  })

  server.app.bsky.actor.getProfiles({
    auth: ctx.authVerifier.moderator,
    handler: async (request) => {
      const res = await ctx.appviewAgent.api.app.bsky.actor.getProfiles(
        request.params,
        await ctx.appviewAuth(ids.AppBskyActorGetProfiles),
      )
      return {
        encoding: 'application/json',
        body: res.data,
      }
    },
  })

  server.app.bsky.feed.getAuthorFeed({
    auth: ctx.authVerifier.moderator,
    handler: async (request) => {
      const res = await ctx.appviewAgent.api.app.bsky.feed.getAuthorFeed(
        request.params,
        await ctx.appviewAuth(ids.AppBskyFeedGetAuthorFeed),
      )
      return {
        encoding: 'application/json',
        body: res.data,
      }
    },
  })

  server.app.bsky.feed.searchPosts({
    auth: ctx.authVerifier.moderator,
    handler: async (request) => {
      const res = await ctx.appviewAgent.api.app.bsky.feed.searchPosts(
        request.params,
        await ctx.appviewAuth(ids.AppBskyFeedSearchPosts),
      )
      return {
        encoding: 'application/json',
        body: res.data,
      }
    },
  })

  server.app.bsky.feed.getPostThread({
    auth: ctx.authVerifier.moderator,
    handler: async (request) => {
      const res = await ctx.appviewAgent.api.app.bsky.feed.getPostThread(
        request.params,
        await ctx.appviewAuth(ids.AppBskyFeedGetPostThread),
      )
      return {
        encoding: 'application/json',
        body: res.data,
      }
    },
  })

  server.app.bsky.feed.getFeedGenerator({
    auth: ctx.authVerifier.moderator,
    handler: async (request) => {
      const res = await ctx.appviewAgent.api.app.bsky.feed.getFeedGenerator(
        request.params,
        await ctx.appviewAuth(ids.AppBskyFeedGetFeedGenerator),
      )
      return {
        encoding: 'application/json',
        body: res.data,
      }
    },
  })

  server.app.bsky.graph.getFollows({
    auth: ctx.authVerifier.moderator,
    handler: async (request) => {
      const res = await ctx.appviewAgent.api.app.bsky.graph.getFollows(
        request.params,
        await ctx.appviewAuth(ids.AppBskyGraphGetFollows),
      )
      return {
        encoding: 'application/json',
        body: res.data,
      }
    },
  })

  server.app.bsky.graph.getFollowers({
    auth: ctx.authVerifier.moderator,
    handler: async (request) => {
      const res = await ctx.appviewAgent.api.app.bsky.graph.getFollowers(
        request.params,
        await ctx.appviewAuth(ids.AppBskyGraphGetFollowers),
      )
      return {
        encoding: 'application/json',
        body: res.data,
      }
    },
  })

  server.app.bsky.graph.getList({
    auth: ctx.authVerifier.moderator,
    handler: async (request) => {
      const res = await ctx.appviewAgent.api.app.bsky.graph.getList(
        request.params,
        await ctx.appviewAuth(ids.AppBskyGraphGetList),
      )
      return {
        encoding: 'application/json',
        body: res.data,
      }
    },
  })

  server.app.bsky.graph.getLists({
    auth: ctx.authVerifier.moderator,
    handler: async (request) => {
      const res = await ctx.appviewAgent.api.app.bsky.graph.getLists(
        request.params,
        await ctx.appviewAuth(ids.AppBskyGraphGetLists),
      )
      return {
        encoding: 'application/json',
        body: res.data,
      }
    },
  })

  server.com.atproto.admin.searchAccounts({
    auth: ctx.authVerifier.moderator,
    handler: async (request) => {
      if (!ctx.pdsAgent) {
        throw new Error('PDS not configured')
      }
      const res = await ctx.pdsAgent.api.com.atproto.admin.searchAccounts(
        request.params,
        await ctx.pdsAuth(ids.ComAtprotoAdminSearchAccounts),
      )
      return {
        encoding: 'application/json',
        body: res.data,
      }
    },
  })

  server.app.bsky.graph.getStarterPack({
    auth: ctx.authVerifier.moderator,
    handler: async (request) => {
      const res = await ctx.appviewAgent.api.app.bsky.graph.getStarterPack(
        request.params,
        await ctx.appviewAuth(ids.AppBskyGraphGetStarterPack),
      )
      return {
        encoding: 'application/json',
        body: res.data,
      }
    },
  })

  server.app.bsky.graph.getStarterPacks({
    auth: ctx.authVerifier.moderator,
    handler: async (request) => {
      const res = await ctx.appviewAgent.api.app.bsky.graph.getStarterPacks(
        request.params,
        await ctx.appviewAuth(ids.AppBskyGraphGetStarterPacks),
      )
      return {
        encoding: 'application/json',
        body: res.data,
      }
    },
  })

  server.app.bsky.graph.getActorStarterPacks({
    auth: ctx.authVerifier.moderator,
    handler: async (request) => {
      const res =
        await ctx.appviewAgent.api.app.bsky.graph.getActorStarterPacks(
          request.params,
          await ctx.appviewAuth(ids.AppBskyGraphGetActorStarterPacks),
        )
      return {
        encoding: 'application/json',
        body: res.data,
      }
    },
  })

  server.app.bsky.feed.getLikes({
    auth: ctx.authVerifier.moderator,
    handler: async (request) => {
      const res = await ctx.appviewAgent.api.app.bsky.feed.getLikes(
        request.params,
        await ctx.appviewAuth(ids.AppBskyFeedGetLikes),
      )
      return {
        encoding: 'application/json',
        body: res.data,
      }
    },
  })

  server.app.bsky.feed.getRepostedBy({
    auth: ctx.authVerifier.moderator,
    handler: async (request) => {
      const res = await ctx.appviewAgent.api.app.bsky.feed.getRepostedBy(
        request.params,
        await ctx.appviewAuth(ids.AppBskyFeedGetRepostedBy),
      )
      return {
        encoding: 'application/json',
        body: res.data,
      }
    },
  })

  server.app.bsky.actor.searchActorsTypeahead({
    auth: ctx.authVerifier.moderator,
    handler: async (request) => {
      const res =
        await ctx.appviewAgent.api.app.bsky.actor.searchActorsTypeahead(
          request.params,
          await ctx.appviewAuth(ids.AppBskyActorSearchActorsTypeahead),
        )
      return {
        encoding: 'application/json',
        body: res.data,
      }
    },
  })
}
