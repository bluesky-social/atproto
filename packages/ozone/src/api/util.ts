import { InvalidRequestError } from '@atproto/xrpc-server'
import { InputSchema as ReportInput } from '../lexicon/types/com/atproto/moderation/createReport'
import {
  REASONOTHER,
  REASONSPAM,
  REASONMISLEADING,
  REASONRUDE,
  REASONSEXUAL,
  REASONVIOLATION,
  REASONAPPEAL,
} from '../lexicon/types/com/atproto/moderation/defs'
import { AccountView } from '../lexicon/types/com/atproto/admin/defs'
import {
  RepoView,
  RepoViewDetail,
  REVIEWCLOSED,
  REVIEWESCALATED,
  REVIEWOPEN,
} from '../lexicon/types/tools/ozone/moderation/defs'
import { ModerationEvent } from '../db/schema/moderation_event'
import { ModerationSubjectStatusRow } from '../mod-service/types'
import AppContext from '../context'
import { Member } from '../db/schema/member'
import {
  ROLEADMIN,
  ROLEMODERATOR,
  ROLETRIAGE,
} from '../lexicon/types/tools/ozone/team/defs'
import { ids } from '../lexicon/lexicons'

export const getPdsAccountInfo = async (
  ctx: AppContext,
  did: string,
): Promise<AccountView | null> => {
  const agent = ctx.pdsAgent
  if (!agent) return null
  const auth = await ctx.pdsAuth(ids.ComAtprotoAdminGetAccountInfo)
  if (!auth) return null
  try {
    const res = await agent.api.com.atproto.admin.getAccountInfo({ did }, auth)
    return res.data
  } catch {
    return null
  }
}

export const addAccountInfoToRepoViewDetail = (
  repoView: RepoViewDetail,
  accountInfo: AccountView | null,
  includeEmail = false,
): RepoViewDetail => {
  if (!accountInfo) return repoView
  return {
    ...repoView,
    email: includeEmail ? accountInfo.email : undefined,
    invitedBy: accountInfo.invitedBy,
    invitesDisabled: accountInfo.invitesDisabled,
    inviteNote: accountInfo.inviteNote,
    invites: accountInfo.invites,
    emailConfirmedAt: accountInfo.emailConfirmedAt,
    deactivatedAt: accountInfo.deactivatedAt,
  }
}

export const addAccountInfoToRepoView = (
  repoView: RepoView,
  accountInfo: AccountView | null,
  includeEmail = false,
): RepoView => {
  if (!accountInfo) return repoView
  return {
    ...repoView,
    email: includeEmail ? accountInfo.email : undefined,
    invitedBy: accountInfo.invitedBy,
    invitesDisabled: accountInfo.invitesDisabled,
    inviteNote: accountInfo.inviteNote,
    deactivatedAt: accountInfo.deactivatedAt,
  }
}

export const getReasonType = (reasonType: ReportInput['reasonType']) => {
  if (reasonTypes.has(reasonType)) {
    return reasonType as NonNullable<ModerationEvent['meta']>['reportType']
  }
  throw new InvalidRequestError('Invalid reason type')
}

export const getEventType = (type: string) => {
  if (eventTypes.has(type)) {
    return type as ModerationEvent['action']
  }
  throw new InvalidRequestError('Invalid event type')
}

export const getReviewState = (reviewState?: string) => {
  if (!reviewState) return undefined
  if (reviewStates.has(reviewState)) {
    return reviewState as ModerationSubjectStatusRow['reviewState']
  }
  throw new InvalidRequestError('Invalid review state')
}

const reviewStates = new Set([REVIEWCLOSED, REVIEWESCALATED, REVIEWOPEN])

const reasonTypes = new Set([
  REASONOTHER,
  REASONSPAM,
  REASONMISLEADING,
  REASONRUDE,
  REASONSEXUAL,
  REASONVIOLATION,
  REASONAPPEAL,
])

const eventTypes = new Set([
  'tools.ozone.moderation.defs#modEventTakedown',
  'tools.ozone.moderation.defs#modEventAcknowledge',
  'tools.ozone.moderation.defs#modEventEscalate',
  'tools.ozone.moderation.defs#modEventComment',
  'tools.ozone.moderation.defs#modEventLabel',
  'tools.ozone.moderation.defs#modEventReport',
  'tools.ozone.moderation.defs#modEventMute',
  'tools.ozone.moderation.defs#modEventUnmute',
  'tools.ozone.moderation.defs#modEventMuteReporter',
  'tools.ozone.moderation.defs#modEventUnmuteReporter',
  'tools.ozone.moderation.defs#modEventReverseTakedown',
  'tools.ozone.moderation.defs#modEventEmail',
  'tools.ozone.moderation.defs#modEventResolveAppeal',
  'tools.ozone.moderation.defs#modEventTag',
  'tools.ozone.moderation.defs#modEventDivert',
])

export const getMemberRole = (role: string) => {
  if (memberRoles.has(role)) {
    return role as Member['role']
  }
  throw new InvalidRequestError('Invalid member role')
}

const memberRoles = new Set([ROLEADMIN, ROLEMODERATOR, ROLETRIAGE])
