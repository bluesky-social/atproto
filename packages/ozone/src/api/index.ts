import { Server } from '../lexicon'
import AppContext from '../context'
import createReport from './moderation/createReport'
import emitModerationEvent from './admin/emitModerationEvent'
import searchRepos from './admin/searchRepos'
import adminGetRecord from './admin/getRecord'
import getRepo from './admin/getRepo'
import queryModerationStatuses from './admin/queryModerationStatuses'
import queryModerationEvents from './admin/queryModerationEvents'
import getModerationEvent from './admin/getModerationEvent'
import queryLabels from './label/queryLabels'
import subscribeLabels from './label/subscribeLabels'
import fetchLabels from './temp/fetchLabels'
import createCommunicationTemplate from './admin/createCommunicationTemplate'
import updateCommunicationTemplate from './admin/updateCommunicationTemplate'
import deleteCommunicationTemplate from './admin/deleteCommunicationTemplate'
import listCommunicationTemplates from './admin/listCommunicationTemplates'
import proxied from './proxied'

export * as health from './health'

export * as wellKnown from './well-known'

export default function (server: Server, ctx: AppContext) {
  createReport(server, ctx)
  emitModerationEvent(server, ctx)
  searchRepos(server, ctx)
  adminGetRecord(server, ctx)
  getRepo(server, ctx)
  getModerationEvent(server, ctx)
  queryModerationEvents(server, ctx)
  queryModerationStatuses(server, ctx)
  queryLabels(server, ctx)
  subscribeLabels(server, ctx)
  fetchLabels(server, ctx)
  listCommunicationTemplates(server, ctx)
  createCommunicationTemplate(server, ctx)
  updateCommunicationTemplate(server, ctx)
  deleteCommunicationTemplate(server, ctx)
  proxied(server, ctx)
  return server
}
