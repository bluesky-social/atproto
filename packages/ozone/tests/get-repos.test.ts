import {
  SeedClient,
  TestNetwork,
  TestOzone,
  basicSeed,
  ModeratorClient,
} from '@atproto/dev-env'
import { AtpAgent } from '@atproto/api'
import {
  REASONOTHER,
  REASONSPAM,
} from '../src/lexicon/types/com/atproto/moderation/defs'
import { forSnapshot } from './_util'
import { ids } from '../src/lexicon/lexicons'

describe('admin get multiple repos', () => {
  let network: TestNetwork
  let ozone: TestOzone
  let agent: AtpAgent
  let pdsAgent: AtpAgent
  let sc: SeedClient
  let modClient: ModeratorClient

  beforeAll(async () => {
    network = await TestNetwork.create({
      dbPostgresSchema: 'ozone_admin_get_repos',
    })
    ozone = network.ozone
    agent = ozone.getClient()
    pdsAgent = network.pds.getClient()
    sc = network.getSeedClient()
    modClient = ozone.getModClient()
    await basicSeed(sc)
    await pdsAgent.com.atproto.server.deactivateAccount(
      {},
      { encoding: 'application/json', headers: sc.getHeaders(sc.dids.dan) },
    )
    await network.processAll()
  })

  afterAll(async () => {
    await network.close()
  })

  beforeAll(async () => {
    await modClient.emitEvent({
      event: { $type: 'tools.ozone.moderation.defs#modEventAcknowledge' },
      subject: {
        $type: 'com.atproto.admin.defs#repoRef',
        did: sc.dids.alice,
      },
    })
    await sc.createReport({
      reportedBy: sc.dids.bob,
      reasonType: REASONSPAM,
      subject: {
        $type: 'com.atproto.admin.defs#repoRef',
        did: sc.dids.alice,
      },
    })
    await sc.createReport({
      reportedBy: sc.dids.carol,
      reasonType: REASONOTHER,
      reason: 'defamation',
      subject: {
        $type: 'com.atproto.admin.defs#repoRef',
        did: sc.dids.alice,
      },
    })
    await modClient.emitEvent({
      event: { $type: 'tools.ozone.moderation.defs#modEventTakedown' },
      subject: {
        $type: 'com.atproto.admin.defs#repoRef',
        did: sc.dids.alice,
      },
    })
  })

  it('gets multiple repos by did', async () => {
    const { data } = await agent.tools.ozone.moderation.getRepos(
      { dids: [sc.dids.alice, sc.dids.bob, 'did:web:xyz'] },
      { headers: await ozone.modHeaders(ids.ToolsOzoneModerationGetRepos) },
    )

    expect(forSnapshot(data)).toMatchSnapshot()
  })
})
