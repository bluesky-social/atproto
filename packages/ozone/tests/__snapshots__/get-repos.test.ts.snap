// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`admin get multiple repos gets multiple repos by did 1`] = `
Object {
  "repos": Array [
    Object {
      "$type": "tools.ozone.moderation.defs#repoViewDetail",
      "did": "user(0)",
      "email": "alice@test.com",
      "handle": "alice.test",
      "indexedAt": "1970-01-01T00:00:00.000Z",
      "invites": Array [],
      "invitesDisabled": false,
      "labels": Array [
        Object {
          "cts": "1970-01-01T00:00:00.000Z",
          "sig": Object {
            "$bytes": "sig(0)",
          },
          "src": "user(2)",
          "uri": "user(0)",
          "val": "!takedown",
          "ver": 1,
        },
      ],
      "moderation": Object {
        "subjectStatus": Object {
          "createdAt": "1970-01-01T00:00:00.000Z",
          "hosting": Object {
            "$type": "tools.ozone.moderation.defs#accountHosting",
            "status": "unknown",
          },
          "id": 1,
          "lastReportedAt": "1970-01-01T00:00:00.000Z",
          "lastReviewedAt": "1970-01-01T00:00:00.000Z",
          "lastReviewedBy": "user(1)",
          "reviewState": "tools.ozone.moderation.defs#reviewClosed",
          "subject": Object {
            "$type": "com.atproto.admin.defs#repoRef",
            "did": "user(0)",
          },
          "subjectBlobCids": Array [],
          "subjectRepoHandle": "alice.test",
          "tags": Array [
            "lang:und",
          ],
          "takendown": true,
          "updatedAt": "1970-01-01T00:00:00.000Z",
        },
      },
      "relatedRecords": Array [
        Object {
          "$type": "app.bsky.actor.profile",
          "avatar": Object {
            "$type": "blob",
            "mimeType": "image/jpeg",
            "ref": Object {
              "$link": "cids(0)",
            },
            "size": 3976,
          },
          "createdAt": "1970-01-01T00:00:00.000Z",
          "description": "its me!",
          "displayName": "ali",
          "labels": Object {
            "$type": "com.atproto.label.defs#selfLabels",
            "values": Array [
              Object {
                "val": "self-label-a",
              },
              Object {
                "val": "self-label-b",
              },
            ],
          },
        },
      ],
    },
    Object {
      "$type": "tools.ozone.moderation.defs#repoViewDetail",
      "did": "user(3)",
      "email": "bob@test.com",
      "handle": "bob.test",
      "indexedAt": "1970-01-01T00:00:00.000Z",
      "invites": Array [],
      "invitesDisabled": false,
      "labels": Array [],
      "moderation": Object {},
      "relatedRecords": Array [
        Object {
          "$type": "app.bsky.actor.profile",
          "avatar": Object {
            "$type": "blob",
            "mimeType": "image/jpeg",
            "ref": Object {
              "$link": "cids(0)",
            },
            "size": 3976,
          },
          "createdAt": "1970-01-01T00:00:00.000Z",
          "description": "hi im bob label_me",
          "displayName": "bobby",
        },
      ],
    },
    Object {
      "$type": "tools.ozone.moderation.defs#repoViewNotFound",
      "did": "did:web:xyz",
    },
  ],
}
`;
